### PR TITLE
Async Connection Layer

### PR DESCRIPTION
**Pull request by IHEII**
### What changes were proposed in this pull request?

- Client Runtime.
- Asynchronous Connections.

### Why are the changes needed?

For better performance.

### Will break the compatibility? How if so?
Users will currently not be aware of asynchronous changes in the connection layer, and they will use the existing code without changing it.

### Does this PR introduce any user-facing change?
Users will currently not be aware of asynchronous changes in the connection layer, and they will use the existing code without changing it.

### How was this patch tested?
3.2.4

### Checklist

- [x] I've run the tests to see all new and existing tests pass.
- [x] If this Pull Request resolves an issue, I linked to the issue in the text above.
- [x] I've informed the technical writer about the documentation change if necessary.

### CODE REVIEW

**Comment by IHEII** [comment]
This MR is the premise of implementing the asynchronous API(#16)

File: src/rpc/mod.rs [hunk 1]
```diff
@@ -355,28 +356,23 @@ impl Connection {
...
-        for (_, sender) in requests.iter() {
+        for (_, sender) in requests.drain() {
             if let Err(e) = sender.send(Err(CommonErr(
                 CommonErrCode::Rpc,
                 "connection reader exits".to_owned(),
             ))) {
                 error!(
                     "Connection::cancel_requests: fail to send cancel message, err:{}",
-                    e
+                    e.err().unwrap().to_string()
```
Comment (ShiKaiWi) [inline src/rpc/mod.rs:372]: Is this unwrap safe? Can we just print `{e:?}`?

File: src/rpc/mod.rs [hunk 2]
```diff
@@ -544,42 +540,45 @@ impl Connection {
         let rx = self.send(req, channel_id)?;
 
         if payload.timeout_millis() == 0 {
-            // no-wait request,return Ok directly
+            // no-wait request,return Ok directly2
```
Comment (ShiKaiWi) [inline src/rpc/mod.rs:543]:
```suggestion
            // no-wait request,return Ok directly
```

File: src/rpc/mod.rs [hunk 3]
```diff
@@ -544,42 +540,45 @@ impl Connection {
...
         }
 
-        let resp = match rx.recv_timeout(timeout) {
-            Ok(resp) => {
-                self.on_recv_in_time();
-                resp.map_err(|e| {
-                    error!(
+        // TODO: remove block_on with rx.await
+        let resp = self.runtimes.default_runtime.block_on(async move {
+            match TokioTimeout(timeout, rx).await {
```
Comment (ShiKaiWi) [inline src/rpc/mod.rs:549]: When use a free function from a module, just use `<module_name>::<function_name>`:
```suggestion
            match tokio::time::timeout(timeout, rx).await {
```

**Comment by WeiXinChan** [comment]
LGTM

File: src/rpc/mod.rs (hunk 2 above)
Comment (WeiXinChan) [inline src/rpc/mod.rs:543]: directly2 -> directly

File: src/client/mod.rs [hunk 4]
```diff
@@ -199,6 +203,10 @@ impl Default for ClientConfig {
             conn_init_thread_num: 2,
             query_concurrency_limit: None,
 
+            conn_reader_threads: 6,
+            conn_writer_threads: 4,
+            default_threads_num: 2,
```
Comment (ShiKaiWi) [inline src/client/mod.rs:208]: Why not keep them in the same style? And how about setting all these runtime's thread number is 2 by default?
```suggestion
            conn_reader_thread_num: 2,
            conn_writer_thread_num: 2,
            default_threads_num: 2,
```

**Review by Rachelint — APPROVED** [review]

File: src/rpc/mod.rs [hunk 5]
```diff
@@ -302,12 +307,13 @@ impl Connection {
         let mut buf = BytesMut::with_capacity(READ_BUF_SIZE);
         loop {
             if let Ok(()) = signal_receiver.try_recv() {
+                error!("DEBUG: out0");
```
Comment (Rachelint) [inline src/rpc/mod.rs:310]: Make it to debug! and refactor the logged message.

File: src/rpc/mod.rs [hunk 6]
```diff
@@ -355,28 +356,23 @@ impl Connection {
...
             }
         }
-        if let Err(err) = read_stream.shutdown(Shutdown::Read) {
-            warn!(
-                "Connection::process_reading_data fail to close read stream to {}, err:{}",
-                addr, err
-            );
-        }
+        // TODO: stream close
+        drop(read_stream);
```
Comment (Rachelint) [inline src/rpc/mod.rs:360]: Why not shutdown like write_stream?

File: src/rpc/mod.rs [hunk 7]
```diff
@@ -138,16 +147,16 @@ impl ConnectionSender {
...
                             break;
                         }
                     }
                 }
 
                 active.store(false, Ordering::Release);
 
-                if let Err(err) = write_stream.shutdown(Shutdown::Write) {
+                if let Err(err) = write_stream.shutdown().await {
                     error!("Fail to close write stream to {}, err:{}", addr, err);
```
[the remark this answers is not in the document]
Comment (IHEII) [reply src/rpc/mod.rs]: done

**Review by IHEII — COMMENTED** [review]
Fix review

File: src/rpc/mod.rs (hunk 6 above)
Comment (Rachelint) [reply src/rpc/mod.rs:363]: No need to drop I think, when the function end, it will be dropped automatically.
In fact this is the classic RAII.

**Review by IHEII — COMMENTED** [review]
Fix review

File: src/client/mod.rs [hunk 8]
```diff
@@ -155,6 +155,12 @@ pub struct ClientConfig {
     pub conn_init_thread_num: usize,
     pub query_concurrency_limit: Option<usize>,
 
+    pub conn_reader_thread_num: usize,
+    pub conn_writer_thread_num: usize,
+    pub default_thread_num: usize,
+
+    pub conn_sender_mpsc_channel: usize,
```
Comment (ShiKaiWi) [inline src/client/mod.rs:162]: Maybe we need a more readable name for this option:
```suggestion
    pub max_reqs_in_flight_per_conn: usize,
```

Comment (IHEII) [reply src/client/mod.rs]: Done, using max_inflight_reqs_per_conn instead.

**Review by ShiKaiWi — APPROVED** [review]
LGTM

**Pull request merged**